src/connector/src/sink/dynamodb.rs

```diff
@@ -0,0 +1,329 @@
+                })
+                .map(|_| ())?
+        }
+        if !self.delete_items.is_empty() {
+            let new_items = std::mem::take(&mut self.delete_items);
+            self.client
+                .delete_item()
+                .table_name(self.table.clone())
+                .set_key(Some(new_items))
+                .send()
+                .await
+                .map_err(|e| {
+                    SinkError::DynamoDb(
+                        anyhow!(e).context("failed to delete item from DynamoDB sink"),
+                    )
+                })
+                .map(|_| ())?
+        }
+
+        Ok(())
+    }
+}
+
+pub struct DynamoDbSinkWriter {
+    payload_writer: DynamoDbPayloadWriter,
+    formatter: DynamoDbFormatter,
+}
+
+impl DynamoDbSinkWriter {
+    pub async fn new(config: DynamoDbConfig, schema: Schema) -> Result<Self> {
+        let client = config.common.build_client().await?;
+
+        let payload_writer = DynamoDbPayloadWriter {
+            insert_items: HashMap::new(),
+            delete_items: HashMap::new(),
+            client,
+            table: config.common.table,
+        };
+
+        Ok(Self {
+            payload_writer,
+            formatter: DynamoDbFormatter { schema },
+        })
+    }
+
+    async fn write_chunk_inner(&mut self, chunk: StreamChunk) -> Result<()> {
+        for (op, row) in chunk.rows() {
+            match op {
+                Op::Insert | Op::UpdateInsert => {
+                    for (k, v) in self.formatter.format_row(row)? {
+                        self.payload_writer.write_one_insert(k, v);
+                    }
+                }
+                Op::Delete | Op::UpdateDelete => {
+                    for (k, v) in self.formatter.format_row(row)? {
+                        self.payload_writer.write_one_delete(k, v);
+                    }
+                }
+            }
+        }
+        self.payload_writer.write_chunk().await
+    }
+}
+
+impl AsyncTruncateSinkWriter for DynamoDbSinkWriter {
+    async fn write_chunk<'a>(
+        &'a mut self,
+        chunk: StreamChunk,
+        _add_future: DeliveryFutureManagerAddFuture<'a, Self::DeliveryFuture>,
+    ) -> Result<()> {
+        self.write_chunk_inner(chunk).await
+    }
+}
+
+struct DynamoDbFormatter {
+    schema: Schema,
+}
+
+impl DynamoDbFormatter {
+    fn format_row(&self, row: RowRef<'_>) -> Result<Vec<(String, AttributeValue)>> {
+        row.iter()
+            .zip_eq_debug((self.schema.clone()).into_fields())
+            .map(|(scalar, field)| {
+                map_data_type(scalar, &field.data_type()).map(|attr| (field.name, attr))
+            })
+            .collect()
+    }
+}
+
+fn map_data_type(
+    scalar_ref: Option<ScalarRefImpl<'_>>,
+    data_type: &DataType,
+) -> Result<AttributeValue> {
+    let Some(scalar_ref) = scalar_ref else {
+        return Ok(AttributeValue::Null(true));
+    };
+    let attr = match (data_type, scalar_ref) {
```

Comment:
It seems that matching the `data_type` only could make code a bit cleaner

Reply:
I'm worried that the `DataType` is different from the `ScalarRefImpl` variant, and just matching `data_type` would mean requiring `scalar_ref.into_foo()` (which could panic in some case). Do you mean that this function `map_data_type` is only used here and does not need to consider possible the tow varant different situations?

Reply:
I think each `DataType` must have only one possible `ScalarRefImpl`, so no need to match `ScalarRefImpl`.